File: backend/user_display.py
"""One way to render a user as a label (#781).

Free-text `owner` is gone from evidence tracking, so every surface that used to
print that string now prints a resolved person instead — the task list, the
task summary, and the generated evidence schedule. Three copies of
``display_name or email`` is how the next inconsistency starts, and this PR's
whole argument is that one concept should have one representation.

Mirrors ``userLabel()`` in ``webclient/src/data/userDisplay.ts``.
"""
from __future__ import annotations

from typing import Optional


def user_label(user) -> Optional[str]:
    """Display label for a resolved user, or None when there is no user.

    Display name where there is one, email otherwise. An email is a worse label
    than a name, and a much better one than a blank cell.
    """
    if user is None:
        return None
    return (getattr(user, "display_name", None) or "").strip() or user.email

File: webclient/src/data/userDisplay.ts
import type { EvidenceTracking, UserSimple } from '../types'

/**
 * How a user is named in the UI.
 *
 * Mirrors `user_label()` in `backend/user_display.py`, which produces the same
 * string for the API payloads that don't serialise a whole user object. Keep
 * the two in step: a name that changes shape depending on which endpoint drew
 * it reads as two different people.
 */
export function userLabel(user: UserSimple): string {
  return user.display_name?.trim() || user.email
}

/**
 * Who is answerable for an evidence row, as a display/grouping label (#781).
 *
 * Until #781 the reporting and dashboard breakdowns grouped on
 * `evidence_tracking.owner` — a free-text "Owner Team" box that no task, no
 * reminder and no queue ever read. The box is gone, so the answer now comes
 * from the columns that actually drive work: the accountable owner, falling
 * back to the assignee when nobody has been made accountable yet.
 *
 * Returns 'Unassigned' rather than '' so it can be used directly as a group
 * key — an empty bucket heading is indistinguishable from a rendering bug.
 */
export function evidenceOwnerLabel(tracking?: EvidenceTracking | null): string {
  const user = tracking?.owner_user || tracking?.assigned_user
  return user ? userLabel(user) : 'Unassigned'
}

/**
 * The user id behind ``evidenceOwnerLabel`` (#822 phase 2).
 *
 * The reporting and dashboard breakdowns group on the LABEL, which is right —
 * it is what the heading has to read. But a label cannot be turned back into a
 * person: two people can share a display name and 'Unassigned' is not anybody.
 * So the id is carried alongside, and a group is only labelled a contractor's
 * when every row in it resolved to the same person.
 *
 * Null when nobody is answerable, which is the ordinary case for untracked
 * evidence and must render as no badge rather than as a missing lookup.
 */
export function evidenceOwnerUserId(tracking?: EvidenceTracking | null): string | null {
  const user = tracking?.owner_user || tracking?.assigned_user
  return user?.id ? String(user.id) : null
}
